drivers/power/max42500/max42500.c

```diff
@@ -0,0 +1,904 @@
+ * @return 0 in case of success, error code otherwise
+ */
+int max42500_reg_write(struct max42500_dev *desc, struct max42500_reg_st reg)
+{
+	uint8_t i2c_data[MAX42500_I2C_WR_FRAME_SIZE] = {0};
+	uint8_t bytes_number;
+
+	bytes_number = (desc->pece) ? (MAX42500_I2C_WR_FRAME_SIZE - 1) : 2;
+	i2c_data[0] = (desc->comm_desc->slave_address << 1);
+	i2c_data[1] = reg.addr;
+	i2c_data[2] = reg.value;
+
+	if (desc->pece)
+		i2c_data[3] = crc8(i2c_data, bytes_number);
+
+	return no_os_i2c_write(desc->comm_desc, &i2c_data[1], bytes_number, 1);
+}
+
+/**
+ * @brief Wrap the write register function to give it a modern signature.
+ * @param desc - The device structure
+ * @param reg - Address of the register to be written.
+ * @param writeval - New value for the register.
+ * @return 0 in case of success, error code otherwise.
+ */
+int max42500_reg_write2(struct max42500_dev *desc,
+			uint32_t reg,
+			uint32_t writeval)
+{
+	desc->regs[reg].value = writeval;
+
+	return max42500_reg_write(desc, desc->regs[reg]);
+}
+
+/**
+ * @brief Set Packet Error Checking enable bit
+ * @param desc - The device structure.
+ * @param pece - The PECE bit value.
+ * @return 0 in case of success, error code otherwise
+ */
+int max42500_set_pece(struct max42500_dev *desc, bool pece)
+{
+	if (!desc)
+		return -EINVAL;
+
+	int ret;
+	struct max42500_reg_st reg = desc->regs[MAX42500_CONFIG1];
+
+	if (pece)
+		reg.value |= MAX42500_CONFIG1_PECE_MASK;
+	else
+		reg.value &= ~MAX42500_CONFIG1_PECE_MASK;
+
+	/* Write register */
+	ret = max42500_reg_write(desc, reg);
+	if (ret)
+		return ret;
+
+	/* Update register value */
+	desc->regs[MAX42500_CONFIG1].value = reg.value;
+
+	return 0;
+}
+
+/**
+ * @brief Set built-in self-test mapping bit
+ * @param desc - The device structure.
+ * @param mbst - The MBST bit value.
+ * @return 0 in case of success, error code otherwise
+ */
+int max42500_set_mbst(struct max42500_dev *desc, bool mbst)
+{
+	if (!desc)
+		return -EINVAL;
+
+	int ret;
+	struct max42500_reg_st reg = desc->regs[MAX42500_CONFIG1];
+
+	if (mbst)
+		reg.value |= MAX42500_CONFIG1_MBST_MASK;
+	else
+		reg.value &= ~MAX42500_CONFIG1_MBST_MASK;
+
+	/* Write register */
+	ret = max42500_reg_write(desc, reg);
+	if (ret)
+		return ret;
+
+	/* Update register value */
+	desc->regs[MAX42500_CONFIG1].value = reg.value;
+
+	return 0;
+}
+
+/**
+ * @brief Set OTP reload reset configuration bit
+ * @param desc - The device structure.
+ * @param rr - The RR bit value.
+ * @return 0 in case of success, error code otherwise
+ */
+int max42500_set_rr(struct max42500_dev *desc, bool rr)
+{
+	if (!desc)
+		return -EINVAL;
+
+	int ret;
+	struct max42500_reg_st reg = desc->regs[MAX42500_CONFIG1];
+
+	if (rr)
+		reg.value |= MAX42500_CONFIG1_RR_MASK;
+	else
+		reg.value &= ~MAX42500_CONFIG1_RR_MASK;
+
+	/* Write register */
+	ret = max42500_reg_write(desc, reg);
+	if (ret)
+		return ret;
+
+	/* Update register value */
+	desc->regs[MAX42500_CONFIG1].value = reg.value;
```

Comment:
I find this `regs` register table where you cache their values makes the code more verbose and also more inefficient and you can also run into issues like `regs` cached version straying from reality.

But in this review I care more about verbosity. For example this register value update can be moved in the `reg_write()` function, this way you simplify all these "feature" r/w routines.

Also, because of how it's designed, you need to kind of write a function for every interesting bit or bitmask in every register resulting again in verbose code. Take [this example](https://github.com/analogdevicesinc/no-OS/blob/main/drivers/adc/ad7606/ad7606.c#L967) where we wrote a `ad7606_spi_write_mask` routine that performs read&write with a mask to any register bit or field, so we don't need a function for every bit and every bit field. Most of our device drivers have such a `write with mask` function.